### PR TITLE
Per-case hard-mining polish: focus capacity on worst-predicted val cases

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The SOTA checkpoint `dc031qpt` makes uneven errors across the 400 training cases — some vehicles are well-predicted, others are systematically hard. **Hard-mining re-weights the per-case loss** to focus capacity on the cases the model currently finds most difficult. By upweighting cases where τ_y/τ_z rel-L2 is highest, the model is forced to improve on its weak spots.

This is motivated directly by kohaku's Phase 1b outlier diagnosis (PR #719): 4 "restored" test cases dominate the test vol_p metric. But even within the normal distribution, there is per-case variance in how well the SOTA predicts. Hard-mining targets that within-distribution variance.

Concretely: at each epoch, compute a **per-case difficulty score** from the prior epoch's validation (or training) rel-L2, then sample batches with probability proportional to the difficulty score. Cases the model gets wrong most are loaded more often. The expected effect is improvement on the worst-predicted cases (likely τ_y/τ_z where the gap is largest) with minimal regression on already-good cases.

### Step 1 — Track per-case running rel-L2

Add an EMA buffer `case_difficulty[N_cases]` initialized to 1.0 (uniform weighting). After each validation pass, update:
```python
for case_id in val_case_ids:
    rel_l2 = per_case_val_rel_l2[case_id]  # from val loop
    case_difficulty[case_id] = alpha * rel_l2 + (1 - alpha) * case_difficulty[case_id]
```
Use `alpha = 0.5` (fast update, responsive to recent epochs).

### Step 2 — Weighted sampling at data-loader level

Expose the difficulty buffer to the DataLoader via `WeightedRandomSampler`. Sampling probability for case `i` is proportional to `case_difficulty[i] ** beta`, where `beta` controls focus sharpness:
- `beta=0.0` → uniform sampling (baseline, no hard-mining)
- `beta=0.5` → gentle upweighting of hard cases (recommended default)
- `beta=1.0` → sampling proportional to difficulty (aggressive)

Add flags `--hardmining-alpha 0.5` (EMA decay) and `--hardmining-beta 0.5` (sharpness).

### Step 3 — Difficulty update source

Use **training-time rel-L2** (computed per batch from the live model, EMA-averaged per case) rather than waiting for the full val pass. This allows the difficulty buffer to update continuously. Log `train/hardmining_gini` (Gini coefficient of the difficulty distribution) to W&B as a diagnostic — if mining is working, Gini should increase over training as weights concentrate on the hard cases.

### Reproduce command (from `/workspace/senpai/target`)

Two-arm sweep: beta=0.5 (Arm A) and beta=1.0 (Arm B), both resuming from SOTA.

```bash
# Arm A — gentle hard-mining (beta=0.5)
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/dc031qpt/checkpoint.pt \
  --agent tanjiro --wandb-group yi-round41-hardmining-polish \
  --wandb-name tanjiro/hardmining-polish-beta05 \
  --learnable-pe --optimizer lion --lr 3e-7 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 0 --use-ema --ema-decay 0.999 --epochs 5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --hardmining-alpha 0.5 --hardmining-beta 0.5

# Arm B — aggressive hard-mining (beta=1.0), launch after Arm A if Arm A passes EP2 gate
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/dc031qpt/checkpoint.pt \
  --agent tanjiro --wandb-group yi-round41-hardmining-polish \
  --wandb-name tanjiro/hardmining-polish-beta10 \
  --learnable-pe --optimizer lion --lr 3e-7 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 0 --use-ema --ema-decay 0.999 --epochs 5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --hardmining-alpha 0.5 --hardmining-beta 1.0
```

dc031qpt artifact download if needed:
```bash
wandb artifact get wandb-applied-ai-team/senpai-v1-drivaerml/model-nezuko-terminal-lr-3e-7-dc031qpt:latest \
  --root /workspace/senpai/target/artifacts/dc031qpt/
```

### Kill gates

- **EP1 Arm A:** val_abupt > 7.50% — kill (mining is hurting the already-good cases)
- **EP2 Arm A:** val_abupt > 7.45% — kill
- **EP2 Arm B:** val_abupt > max(Arm A EP2 val_abupt) + 0.05pp — kill Arm B (aggressive mining is worse than gentle)

### Diagnostics to watch

- `train/hardmining_gini` — should increase over epochs as weighting concentrates (if it stays at 0.0 the implementation is broken)
- `val/wall_shear_y_rel_l2_pct` and `val/wall_shear_z_rel_l2_pct` — primary beneficiaries if mining helps
- Arm A vs Arm B comparison: does more aggressive mining (higher beta) help or hurt?

## Baseline

**Current SOTA (merge bar):** val_abupt = **7.3767%**, test_abupt = **8.7015%** (PR #681 nezuko, W&B run `dc031qpt`)

Per-axis val/test SOTA:
| Metric | val | test | AB-UPT |
|---|---|---|---|
| abupt_axis_mean | 7.3767% | 8.7015% | — |
| surface_p | 4.8515% | 4.6236% | 3.82% |
| τ_y | 9.5832% | 9.5964% | 3.65% |
| τ_z | 11.0377% | 10.7383% | 3.63% |
| vol_p | 4.3066% | 11.3738% | 6.08% |

## What to report

Per-channel val/test breakdown for Arm A and Arm B at best-by-val-abupt checkpoint:

| Arm | beta | val_abupt | test_abupt | τ_y val | τ_z val | sp val | vp val | gini_final |
|---|---|---|---|---|---|---|---|---|

Plus: which cases had the highest difficulty weight at training end? (Log top-10 most-sampled case IDs and their difficulty scores.) This tells us whether the mining focused on structurally interesting cases or random noise.

Post terminal `SENPAI-RESULT` with val + test for the best arm.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT tanjiro:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["o37boo8p","9yzramki"],"primary_metric":{"name":"full_val_primary/abupt_axis_mean_rel_l2_pct","value":7.3731},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":8.7012}}

## Results

Both arms ran 5-epoch hard-mining polish on top of SOTA checkpoint `dc031qpt` and completed cleanly. Best epoch was EP1 in both arms — EP2/EP3 drifted slightly worse, consistent with terminal-LR polish on a near-converged model. Wall-clock: Arm A = 155.6 min (2 epochs reached, run ended), Arm B = 275.6 min (3 epochs reached, run ended). EP1 was a slow startup epoch (~117 min) due to compile/warm-up, EP2+ ran in ~20–25 min each.

### Headline table (best-by-val checkpoint, full_val + test)

| Arm | beta | val_abupt | test_abupt | τ_y val | τ_z val | sp val | vp val | gini_final |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| SOTA `dc031qpt` (PR #681) | — | 7.3767% | 8.7015% | 9.5832% | 11.0377% | 4.8515% | 4.3066% | — |
| **A** | 0.5 | **7.3731%** | **8.7012%** | 9.5743% | 11.0327% | 4.8515% | 4.3020% | 0.1912 |
| **B** | 1.0 | **7.3716%** | 8.7025% | 9.5757% | 11.0306% | 4.8499% | 4.2995% | 0.1891 |

Δ vs SOTA:

| Arm | val_abupt | test_abupt | τ_y val | τ_z val | sp val | vp val |
|---|---:|---:|---:|---:|---:|---:|
| A (β=0.5) | −0.0036pp | **−0.0003pp** | −0.0089pp | −0.0050pp | 0.0000pp | −0.0046pp |
| B (β=1.0) | **−0.0051pp** | +0.0010pp | −0.0075pp | −0.0071pp | −0.0016pp | −0.0071pp |

**Best by val:** Arm B (β=1.0) — but the gain (−0.0051pp) is well within run-to-run noise and worsens test by +0.0010pp.
**Best by test (the paper-facing metric):** Arm A (β=0.5) — but the gain (−0.0003pp) is essentially tied with SOTA.

### Per-channel test breakdown

| Metric | SOTA | Arm A | Arm B | AB-UPT |
|---|---:|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | 8.7015% | **8.7012%** | 8.7025% | — |
| `test_primary/surface_pressure_rel_l2_pct` | 4.6236% | **4.6229%** | 4.6246% | 3.82% |
| `test_primary/wall_shear_rel_l2_pct` | 8.3214% | **8.3191%** | 8.3225% | 7.29% |
| `test_primary/wall_shear_x_rel_l2_pct` | 7.1753% | **7.1746%** | 7.1775% | 5.35% |
| `test_primary/wall_shear_y_rel_l2_pct` | 9.5964% | **9.5922%** | 9.5939% | 3.65% |
| `test_primary/wall_shear_z_rel_l2_pct` | 10.7383% | **10.7345%** | 10.7405% | 3.63% |
| `test_primary/volume_pressure_rel_l2_pct` | 11.3738% | 11.3816% | 11.3758% | 6.08% |

Arm A wins every per-channel test metric except `volume_pressure` (where Arm B is marginally better, but both regress slightly vs SOTA on test_vp by +0.002–0.008pp).

### Hard-mining diagnostics

- **Final Gini** ≈ 0.19 in both arms — mining concentrated weight, but the distribution stabilized fast (per-epoch Gini values: Arm A `0.1920 → 0.1912`; Arm B `0.1914 → 0.1924 → 0.1891`).
- **Top-6 hardest cases identified consistently across both arms** (difficulty ≈ 0.32–0.35; rest of the train set is < 0.13):
  - `run_44`, `run_184`, `run_249`, `run_310`, `run_416`, `run_484`
- These six are likely structurally distinct (plausibly the "restored" public cases flagged in kohaku's PR #719 outlier diagnosis).
- Arm B (β=1.0) and Arm A (β=0.5) produced near-identical Gini and top-cases lists — the choice of β is mostly cosmetic at this difficulty distribution because the same 6 cases dominate at both sharpness levels.

### Validation trajectory (best-epoch=EP1 in both arms)

| Epoch | Arm A val_abupt | Arm B val_abupt |
|---:|---:|---:|
| 1 | **7.3731%** | **7.3716%** |
| 2 | 7.3745% | 7.3822% |
| 3 | — | 7.3819% |

Kill gates were never close to firing (EP1 7.5% → both well below; EP2 7.45% → both well below).

### Reproduce command

```bash
# Arm A (best-by-test)
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/dc031qpt/checkpoint.pt \
  --agent tanjiro --wandb-group yi-round41-hardmining-polish \
  --wandb-name tanjiro/hardmining-polish-beta05 \
  --learnable-pe --optimizer lion --lr 3e-7 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 0 --use-ema --ema-decay 0.999 --epochs 5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --hardmining-alpha 0.5 --hardmining-beta 0.5

# Arm B (β=1.0): same as above with --hardmining-beta 1.0 and --wandb-name tanjiro/hardmining-polish-beta10
```

Peak GPU memory: 68.1 GB / 96 GB per rank for both arms.

W&B runs:
- Arm A: `o37boo8p` (group `yi-round41-hardmining-polish`, name `tanjiro/hardmining-polish-beta05`)
- Arm B: `9yzramki` (group `yi-round41-hardmining-polish`, name `tanjiro/hardmining-polish-beta10`)

### What happened

The mining mechanism worked as designed — Gini > 0, the same 6 outlier cases were correctly identified across both arms, and weight concentration emerged within EP1. However, the metric impact was **marginal at best, indistinguishable from noise**:

- **Arm A test_abupt = 8.7012%** vs SOTA 8.7015% — a 0.0003pp gain (≈3.4 × 10⁻⁵ relative).
- **Arm B val_abupt = 7.3716%** vs SOTA 7.3767% — a 0.0051pp gain on val that does not transfer to test.

The likely reasons hard-mining did not unlock more at this stage:

1. **lr=3e-7 is too small to redirect capacity meaningfully.** Even with mining, the optimizer can only make ~10⁻⁷-scale weight changes per step, so 5 epochs of mining-weighted gradients cannot move the model far from the resumed SOTA.
2. **Gini saturated immediately.** With α=0.5 EMA and 6 dominant outliers, the difficulty distribution reached steady state in EP1, so EP2+ training was essentially constant-weighted on the same hard cases.
3. **Train hard-cases ≠ val hard-cases.** Mining is computed on the train split; the val split is a different 34 cases. Capacity refocused onto train-hard cases (run_44 etc.) does not necessarily help val cases unless they share structure.
4. **β has near-zero effect** because the difficulty distribution is bimodal (6 cases at ~0.33, the rest at < 0.13). Both β=0.5 and β=1.0 push samples toward the same 6 cases — Arm A and Arm B Ginis are effectively identical (0.1912 vs 0.1891).

Net: the hypothesis "hard-mining at terminal LR moves the SOTA needle" is **falsified for this configuration**. Mining works mechanically but produces a noise-level gain that does not survive train→val→test transfer.

### Suggested follow-ups

1. **Mine on val rel-L2 instead of train rel-L2.** The current implementation observes train-time rel-L2, but the merge bar is val_abupt. Updating the difficulty buffer from per-case val rel-L2 (computed at each `--validation-every`) targets the right distribution. Cost: 1 val pass per update tick, but `--validation-every 1` already pays this.
2. **Hard-mining is more likely to help at higher LR.** A repeat at lr=1e-5 or 5e-6 (one notch above PR #658's lr=5e-6 SWA stage) with 5–10 epochs of mining-weighted updates from `vzprvtaw` (PR #637 SOTA, val_abupt=7.5373%) would give the optimizer real budget to redirect capacity onto the hard cases. This is the Goldilocks zone where mining might compound with the existing wins.
3. **Test-time per-case analysis.** The 6 mined-hard train cases (`run_44`, `run_184`, `run_249`, `run_310`, `run_416`, `run_484`) are likely outliers in train space. A diagnostic comparing their per-channel rel-L2 to the train mean would tell us *why* they are hard (e.g. unusual geometry, tail in target distribution). If they share structure with kohaku's PR #719 4 restored test outliers, a dataset-level intervention (data augmentation, custom feature for those geometries) is more leverage than mining.
4. **Skip the bimodal-flattening β scan; try inverse temperature τ on log-difficulty instead.** Because the difficulty distribution is bimodal, β just toggles between "still mostly the 6 outliers" and "even more the 6 outliers." A softmax over `log(difficulty) / τ` with small τ would let us actually probe the boundary between mining the 6 outliers and mining the next-tier hard cases (run_134, run_492, run_287, run_317).
5. **Composition with τ_y/τ_z loss-axis upweighting.** The two largest absolute gaps to AB-UPT remain `wall_shear_y` (9.59 vs 3.65, 2.6×) and `wall_shear_z` (10.73 vs 3.63, 3.0×). A future round combining hard-mining (per-case axis-targeted) with explicit τ_y/τ_z surface-loss upweighting could attack the dominant remaining gaps directly.

**Recommendation:** Do not merge. The Arm A result is statistically indistinguishable from SOTA on test (−0.0003pp) and the mechanism cost ~7.2 GPU-hours for both arms. The hard-mining infrastructure (sampler, EMA buffer, Gini diagnostics, top-cases table, n_eff metrics) is a reasonable tool to keep around for future rounds, but the polish-stage result here does not justify a new merge bar.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #744 because ADVISOR: Closing — hard-mining does not beat the current yi SOTA.

## Results Summary

| Arm | β | val_abupt | test_abupt |
|-----|---|-----------|------------|
| A | 0.5 | 7.3731% | 8.7012% |
| B | 1.0 | 7.3716% | 8.7025% |
| **yi SOTA (#743)** | — | **7.2733%** | — |

Both arms are 0.098–0.099pp above the yi SOTA. The Δtest vs SOTA is noise-level (< 0.001pp). The mechanism is falsified at terminal LR: hard-mining finds difficulty signal (Gini ~0.19, 6 outlier cases consistently identified) but that signal does not transfer to val/test improvement.

## Why it didn't work

The bimodal difficulty distribution is the core issue. 6 outlier cases (run_44, run_184, run_249, run_310, run_416, run_484) carry ~0.33 difficulty while the rest of the train set sits below 0.13. The sampler concentrates on those 6 cases heavily, but those cases are already learned — they are simply structurally harder geometries. More passes over them does not reduce their loss further. The β parameter is near-irrelevant given this bimodal structure.

## Follow-up ideas (from your own analysis — great observations)

1. **Val-based difficulty buffer**: Compute per-case difficulty on the validation set rather than training loss. Surface cases where val degradation persists across epochs. This is a cleaner signal for what the model truly can't generalize on.
2. **Higher-LR hard-mining**: Run mining at a higher LR (e.g. 5e-4 warm restart) so the model actually has gradient signal to improve on the hard cases it selects.
3. **Per-case test analysis**: Profile the 6 outlier cases against the test set — if they are geometrically OOD from the test distribution, no amount of upweighting will help val/test scores.
4. **Softmax temperature on log-difficulty**: Rather than EMA difficulty → weight, use exp(difficulty / T) with T as a schedule — sharp early, soft late.
5. **Composition with τ_y/τ_z upweighting**: Hard-mining addresses case-level difficulty; annealed wallshear weighting addresses channel-level difficulty. Composing them (mine for cases where τ_z is worst) is worth trying once the wallshear annealing direction is further developed.

Good work on the diagnostic logging (Gini coefficient, difficulty distribution plots). This kind of instrumentation is exactly what makes null results useful..
